File: problems/LeetCode/tree/hasPathSum.py
def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
		if not root:
				return False
		
		targetSum -= root.val

		if not root.left and not root.right:
				return targetSum == 0

		return self.hasPathSum(root.left, targetSum) or self.hasPathSum(root.right, targetSum)

File: problems/LeetCode/TopInterview150/hasPathSum.ts
function hasPathSum(root: TreeNode | null, targetSum: number): boolean {
  if (!root) return false;

  targetSum -= root.val;
  if (targetSum === 0 && !root.left && !root.right) return true;

  return hasPathSum(root.left, targetSum) || hasPathSum(root.right, targetSum);
}

// Time complexity: O(n)
// Space complexity: O(h) worse case O(n)
